cardano-crypto-class/src/Cardano/Crypto/DirectSerialise.hs

```diff
@@ -0,0 +1,136 @@
+import Control.Monad.Class.MonadST (MonadST, stToIO)
+import Data.STRef (newSTRef, readSTRef, writeSTRef)
+import Cardano.Crypto.Libsodium.Memory (copyMem)
+
+-- | Direct deserialization from raw memory.
+--
+-- @directDeserialise f@ should allocate a new value of type 'a', and
+-- call @f@ with a pointer to the raw memory to be filled. @f@ may be called
+-- multiple times, for data structures that store their data in multiple
+-- non-contiguous blocks of memory.
+--
+-- The order in which memory blocks are visited matters.
+class DirectDeserialise a where
+  directDeserialise :: (MonadST m, MonadThrow m) => (Ptr CChar -> CSize -> m ()) -> m a
+
+-- | Direct serialization to raw memory.
+--
+-- @directSerialise f x@ should call @f@ to expose the raw memory underyling
+-- @x@. For data types that store their data in multiple non-contiguous blocks
+-- of memory, @f@ may be called multiple times, once for each block.
+--
+-- The order in which memory blocks are visited matters.
+class DirectSerialise a where
+  directSerialise :: (MonadST m, MonadThrow m) => (Ptr CChar -> CSize -> m ()) -> a -> m ()
+
+-- | Helper function for size-checked serialization.
+-- Verifies that no more than the maximum number of bytes are written, and
+-- returns the actual number of bytes written.
+directSerialiseChecked :: forall m a.
+                          DirectSerialise a
+                       => MonadST m
+                       => MonadThrow m
+                       => (Ptr CChar -> CSize -> m ())
+                       -> Int
+                       -> a
+                       -> m Int
+directSerialiseChecked writeBytes dstsize val = do
+  posRef <- stToIO $ newSTRef 0
+  let pusher :: Ptr CChar -> CSize -> m ()
+      pusher src srcsize = do
+        pos <- stToIO $ readSTRef posRef
+        let pos' = pos + fromIntegral srcsize
+        when (pos' > dstsize) (error "Buffer overrun")
+        writeBytes src (fromIntegral srcsize)
+        stToIO $ writeSTRef posRef pos'
+  directSerialise pusher val
+  stToIO $ readSTRef posRef
+
+-- | Helper function for size-checked deserialization.
+-- Verifies that no more than the maximum number of bytes are read, and returns
+-- the actual number of bytes read.
+directDeserialiseChecked :: forall m a.
+                            DirectDeserialise a
+                         => MonadST m
+                         => MonadThrow m
+                         => (Ptr CChar -> CSize -> m ())
+                         -> Int
+                         -> m (a, Int)
+directDeserialiseChecked readBytes srcsize = do
+  posRef <- stToIO $ newSTRef 0
+  let puller :: Ptr CChar -> CSize -> m ()
+      puller dst dstsize = do
+        pos <- stToIO $ readSTRef posRef
+        let pos' = pos + fromIntegral dstsize
+        when (pos' > srcsize) (error "Buffer overrun")
+        readBytes dst (fromIntegral dstsize)
+        stToIO $ writeSTRef posRef pos'
+  (,) <$> directDeserialise puller <*> stToIO (readSTRef posRef)
```

Comment:
This is a common mistake that we need to either make the user be aware of or even better provide a function that ensures this check (probably even here, but I am not aware of all use cases you have planned for it):

**Check that input has been fully consumed!**

We do have the total size and we do have the total number of bytes consumed. At the end of this action they really should match up, otherwise the target pointer will not be fully filled, thus making the contents non-deterministic, but most importantly it can become a source of bugs and even vulnerabilities when decoding input is bigger than expected. We actually had bugs like that in ledger in the past. Very nasty stuff.

Reply:
Good point actually. I've added two flavors of these functions: one that returns the byte count, leaving the check to the caller, and one that does the check itself. We could skip the former, but I figured that it's probably cleaner to implement one in terms of the other anyway, and with that separation in place, we might as well expose both.